tests/serac_sets.cpp

```diff
@@ -0,0 +1,296 @@
+  serac::FiniteElementState density(*pmesh, FiniteElementState::Options{.order      = 1,
+                                                                        .vector_dim = 1,
+                                                                        .coll       = {},
+                                                                        .ordering   = mfem::Ordering::byVDIM,
+                                                                        .name       = "density",
+                                                                        .alloc_gf   = true});
+
+  // First define PWConstantCoefficient that corresponds to red=1, white = 2, blue = 3
+  mfem::Vector elem_densities(3);
+  elem_densities[0] = 1.;
+  elem_densities[1] = 0.0;
+  elem_densities[2] = 0.0;
+  auto density_coef = mfem::PWConstCoefficient(elem_densities);
+
+  // integrate on the boundary of the bottom
+  mfem::ParFiniteElementSpace& fes = *density.gridFunc().ParFESpace();
+  mfem::ParGridFunction        one_gf(&fes);
+  one_gf = 1.;
+
+  mfem::Array<int> bottom_marker(3);
+  bottom_marker    = 0;
+  bottom_marker[1] = 1;
+
+  auto surface_density_coef = mfem_ext::SurfaceElementAttrCoefficient(*pmesh, density_coef);
+
+  /*
+    Without SurfaceElementAttrCoefficient, the following integration routine will
+    look at the bottom_marker, which tells mfem to integrate on attr = 2 on the boundary.
+    However, density_coef will evaluate the bdrElementTransformation(attr = 2) which evaluates
+    elem_densities[1], which is 0. This is not what we would expect in integrating the density along a boundary.
+
+    So SurfaceElementAttrCoefficient allows you to specify attr = 2 bdr element integration.
+    However, instead of passing the bdrElementTransformation(attr = 2), it passes the
+    ElementTransformation associated with bdrElementTransformation(attr = 2) which is elem attr=1 "red".
+    Thus, density_coef will evaluate as elem_attr=1 and yield the expected answer.
+  */
+
+  mfem::ParLinearForm bottom_wrong(&fes);
+  bottom_wrong.AddBoundaryIntegrator(new mfem::BoundaryLFIntegrator(density_coef), bottom_marker);
+  bottom_wrong.Assemble();
+  double integrate_bottom_wrong = bottom_wrong(one_gf);
+  EXPECT_EQ(0, integrate_bottom_wrong);
+
+  mfem::ParLinearForm bottom(&fes);
+  bottom.AddBoundaryIntegrator(new mfem::BoundaryLFIntegrator(surface_density_coef), bottom_marker);
+  bottom.Assemble();
+  double integrate_bottom = bottom(one_gf);
+  EXPECT_NEAR(1.5, integrate_bottom, 5.e-2);
+
+  mfem::Array<int> right_marker(3);
+  right_marker    = 0;
+  right_marker[2] = 1;
+
+  mfem::ParLinearForm right(&fes);
+  right.AddBoundaryIntegrator(new mfem::BoundaryLFIntegrator(surface_density_coef), right_marker);
+  right.Assemble();
+  double integrate_right = right(one_gf);
+  EXPECT_NEAR(6. / 11., integrate_right, 5.e-2);
+
+  // Reverse attributes: white = 1, red = 2, blue = 3 and integrate on right side
+  std::vector<int> orig_elem_attr = flag.toList();
```

Comment:
Here `orig_elem_attr` is a dense vector list of attributes from indices 0 to `N` element no. We can use `std::transform` to apply a lambda on the dense list to swap _red_ with _white_. The example then shows how we can "hot swap" the reversed red/white coefficient using the `mfem::Integrator` infrastructure and perform an integration on the boundary.

In the move away from coefficients, I think it would be helpful to have an abstraction like `Set` that handles `indices <-> attribute` mappings. `Functional` could then take in a `Set` as an argument, and you would integration over certain attribute/index pairs directly instead of masking. It also might be useful to distinguish between `domain` attributes, `boundary` attributes, and maybe `quadrature`-point attributes using something like types. Based on boundary <-> domain mappings you might be able to make it easier to integrate boundary indices associated with domain indices behind the scenes and avoid having to write something like `SurfaceElementAttrCoefficient`.

Reply:
The abstract `Set` type could be useful when launching material kernels where the material properties are batched (GPU or ROM-based elements that benefit from caching). Since `Set` really is just an abstraction layer for `(index <-> attribute)` pairwise mappings and operations on those, the information can be reused in a lot of different ways.

Some of these operations could be used in determining the actual _consistent_ indexing of domain and boundary d.o.f.s for example and is especially useful if you want to assemble on the GPU.